packages/xrpl/src/models/transactions/payment.ts

```diff
@@ -166,6 +166,24 @@ export interface PaymentMetadata extends TransactionMetadataBase {
 export function validatePayment(tx: Record<string, unknown>): void {
   validateBaseTransaction(tx)
 
+  if (tx.Amount == null) {
```

Comment:
This feels like it's changing the remit of the `validate` functions, which should only be throwing errors and not mutating the objects.

Reply:
Hmm, I agree. But I couldn't find another appropriate place to sneak in this mutation.
Perhaps I could include this change inside `ripple-binary-codec`. But since this is an RPC-related change and has nothing to do with protocol level serialization/de-serialization, it doesn't feel appropriate.

I'm also questioning the necessity of this change. rippled performs validation checks on the `Amount` and `DeliverMax` fields in the `Payment` transactions. So even if the client libraries send incorrect input, the user should get a `tem...` error code from rippled.

Reply:
I think having such a check in `autofill` would be nice.

Yes you get a `tem` code, but it's better to fail fast. Otherwise, why even bother to have the `validate` functions in the first place?

Reply:
I believe you are referring to this `autofill` : https://github.com/XRPLF/xrpl.js/blob/8b596a6687ff89578973429f2b05d622a77b4249/packages/xrpl/src/client/index.ts#L642

It accepts only `SubmittableTransaction`'s. JSON-string representation of certain `Payment` transactions, which include `DeliverMax` field are not serialization friendly with `encode` and `decode` JS functions.


Please refer to this branch, it has two unit tests demonstrating this issue: https://github.com/XRPLF/xrpl.js/commit/d262da9dc4fbd7bc6928b5a7a952ea674429b0f9

If we want to insert this change in `autofill`, we will need to remove Type restrictions of `SubmittableTransaction` and perform cosmetic changes prior to the call to `decode` here: https://github.com/XRPLF/xrpl.js/blob/8b596a6687ff89578973429f2b05d622a77b4249/packages/xrpl/src/sugar/submit.ts#L248

Reply:
If you're using TS, it'll get caught anyways, so that's not who we're worried about.

Reply:
what do you mean? I don't understand.

Irrespective of TS or JS, since the input is a JSON string representation of tx_json, we will need to (1) either update the `decode` function, or (2) modify `tx_json` before and after the decode/encode function calls.

Reply:
`encode`/`decode` are never called in `autofill`, I'm not sure what your concern is.

Reply:
I can't invoke `autofill` directly on a tx_json, because autofill accepts only a `SubmittableTransaction`. A JSON input with `DeliverMax` does not fit the requirements of a `SubmittableTransaction`. So `autofill` cannot be used for our purposes.

Reply:
I need to throw an error if `DeliverMax` != `Amount`. `autofill`'s API does not allow for this possibility

Reply:
I'm exploring the alternative `submit` function: it could be used for populating the `DeliverMax` field. Here is a typical function call hierarchy:
`submit` -> `getSignedTx` -> `autofill`. This is a viable approach. But throwing the above-mentioned error is still not possible with this approach.

what do you think?

Reply:
That's what I'm referring to on TS vs JS. TS will error if you have a `DeliverMax` field, JS will let you continue because it has no type checking. I think an `as any` is fine for this check.

Reply:
A lot of folks call `autofill`/`sign`/`submit` on their own, so updating `submit` wouldn't do anything.

Reply:
Okay, I have implemented this idea here: https://github.com/ckeshava/xrpl.js/pull/new/autofill-temp

But we are losing transaction model checks, as shown by the last unit test -- https://github.com/XRPLF/xrpl.js/blob/13125564baafcc013c9f911bd647ab577c80608f/packages/xrpl/test/client/autofill.test.ts#L165

Is this what you had in mind?

Reply:
What I'm suggesting is that only lines 672-696 use an `as any` version of the object. The rest of the checks can use the normal transaction.

Reply:
Okay, here is another PR: https://github.com/XRPLF/xrpl.js/pull/2689